test/sync/src/androidTest/kotlin/io/realm/test/mongodb/shared/SyncSessionTests.kt

```diff
@@ -110,6 +121,185 @@ class SyncSessionTests {
+
+        val realm1 = Realm.open(config1)
+        val realm2 = Realm.open(config2)
+
+        try {
+            realm1.write {
+                for (i in 0 until 10) {
+                    copyToRealm(
+                        ParentPk().apply {
+                            _id = i.toString()
+                        }
+                    )
+                }
+            }
+            assertEquals(10, realm1.query<ParentPk>().count().find())
+            assertEquals(0, realm2.query<ParentPk>().count().find())
+            assertTrue(realm1.syncSession.uploadAllLocalChanges())
+
+            // Due to the Server Translator, there is a small delay between data
+            // being uploaded and it not being immediately ready for download
+            // on another Realm. In order to reduce the flakyness, we are
+            // re-evaluating the assertion multiple times.
+            for (i in 4 downTo 0) {
+                assertTrue(realm2.syncSession.downloadAllServerChanges())
+                val size = realm2.query<ParentPk>().count().find()
+                when (size == 10L) {
+                    true -> break // Test succeeded
+                    false -> {
+                        if (i == 0) {
+                            throw kotlin.AssertionError("Realm failed to receive download data: $size")
+                        }
+                        if (size != 0L) {
+                            throw AssertionError("Unexpected size: $size")
+                        }
+                        delay(100)
+                    }
+                }
+            }
+        } finally {
+            realm1.close()
+            realm2.close()
+        }
+    }
+
+    @Test
+    fun uploadAndDownload_throwsInsideSyncErrorHandler() = runBlocking {
+        val channel = Channel<SyncSession>(1)
+        var wrongSchemaRealm: Realm? = null
+        val job = async {
+
+            // Create server side Realm with one schema
+            var config = SyncConfiguration.Builder(user, DEFAULT_PARTITION_VALUE, schema = setOf(ParentPk::class, ChildPk::class))
+                .directory(tmpDir)
+                .build()
+            val realm = Realm.open(config)
+            realm.syncSession.uploadAllLocalChanges()
+            realm.close()
+
+            // Create same Realm with another schema, which will cause a Client Reset.
+            config = SyncConfiguration.Builder(user, DEFAULT_PARTITION_VALUE, schema = setOf(ParentPk::class, io.realm.entities.sync.bogus.ChildPk::class))
+                .directory(tmpDir)
+                .name("new_realm.realm")
+                .errorHandler(object : SyncSession.ErrorHandler {
+                    override fun onError(session: SyncSession, error: SyncException) {
+                        channel.trySend(session)
+                    }
+                })
+                .build()
+            wrongSchemaRealm = Realm.open(config)
+        }
+        val session = channel.receive()
```

Comment:
Is this actually executed inside the error callback? I guess it seems so if the below asserts succeeds, but I wouldn't  expect that sending something on a buffered channel would mean the that the receiving end will handle it synchronously in receive 🤔

Reply:
It isn't executed inside the callback as such. But the callback is the only place where the SyncSession is constructed in a special manner, so this test just ensure that if _that_ instance is called, some functionality is not working. I can add a comment about this or find a better name for the test if that would help?

Reply:
Yeah. I guess the test should just have another name them.